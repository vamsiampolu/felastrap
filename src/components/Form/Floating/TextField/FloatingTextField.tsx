import { useFela } from "react-fela";
import type { DivProps, StyleProps } from "../../../common";
import { FloatingTextContext } from "./FloatingTextContext";
import { useFloatingTextField } from "./useFloatingTextField";

type RuleProps = StyleProps<{ group: boolean }>;
export type FloatingTextFieldProps = DivProps & {
  id: string;
  text?: string;
  disabled?: boolean;
  readOnly?: boolean;
  plainText?: boolean;
  value?: string;
};

const rule = () => ({
  display: "block",
  position: "relative",
  marginBottom: "1rem !important",
});

const groupRule = ({ group }: RuleProps) =>
  group
    ? {
        position: "relative",
        flex: "1 1 auto",
        width: "1%",
        minWidth: 0,
        ":focus-within": {
          zIndex: 5,
        },
      }
    : {};

export const FloatingTextField = ({
  className = "",
  children,
  disabled = false,
  readOnly = false,
  plainText = false,
}: FloatingTextFieldProps) => {
  const value = useFloatingTextField({ disabled, readOnly, plainText });
  const { css } = useFela({ group: value.group });
  const cx = `${css([rule, groupRule])} form-floating ${className}`.trim();

  return (
    <div data-testid="floating-text-field" className={cx}>
      <FloatingTextContext.Provider value={value}>
        {children}
      </FloatingTextContext.Provider>
    </div>
  );
};
