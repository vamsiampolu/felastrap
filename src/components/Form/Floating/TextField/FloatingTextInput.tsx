import { useFela } from "react-fela";
import type { TextInputProps } from "../../TextField/TextInput";
import {
  useState,
  type FocusEvent,
  type FocusEventHandler,
  type ForwardedRef,
  ChangeEvent,
} from "react";
import { useFloatingTextContext } from "./useFloatingTextContext";
import { FloatingTextContextValue } from "./FloatingTextContext";
import { Size, StyleProps } from "../../../common";

type RuleProps = StyleProps<{
  group: boolean;
  size: Size;
  readOnly: boolean;
  plainText: boolean;
}>;

const rule = () => ({
  appearance: "none",
  backgroundClip: "padding-box",
  backgroundColor: "#fff",
  border: "1px solid #dee2e6",
  borderRadius: "0.375rem",
  color: "#212529",
  display: "block",
  fontSize: "1rem",
  fontWeight: 400,
  height: "calc(3.5rem + 2px)",
  lineHeight: 1.25,
  minHeight: "calc(3.5rem + 2px)",
  padding: "1rem 0.75rem",
  transition: "border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out",
  width: "100%",
  "@media (prefers-reduced-motion: reduce)": {
    transition: "none",
  },
  ":disabled": {
    backgroundColor: "#e9ecef",
    opacity: 1,
  },
  ":focus": {
    backgroundColor: "#fff",
    borderColor: "#86b7fe",
    boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
    color: "#212529",
    outline: 0,
  },
  "::placeholder": {
    opacity: 1,
    color: "transparent",
  },
  ":not(:placeholder-shown)": {
    paddingTop: "1.625rem",
    paddingBottom: "0.625rem",
  },
  ":-webkit-autofill": {
    paddingTop: "1.625rem",
    paddingBottom: "0.625rem",
  },
});

const groupRule = ({ group }: RuleProps) =>
  group
    ? {
        position: "relative",
        flex: "1 1 auto",
        // width: "1%",
        minWidth: 0,
        ":focus": {
          zIndex: 5,
        },
      }
    : {};

const largeGroupInputRule = ({ group, size }: RuleProps) =>
  group && size === "large"
    ? {
        padding: "0.5rem 1rem",
        fontSize: "1.25rem",
        borderRadius: "0.5rem",
      }
    : {};

const smallGroupInputRule = ({ group, size }: RuleProps) =>
  group && size === "small"
    ? {
        padding: "0.25rem 0.5rem",
        fontSize: "0.875rem",
        borderRadius: "0.25rem",
      }
    : {};

const smallInputRule = ({ size }: RuleProps) =>
  size === "small"
    ? {
        minHeight: "calc(1.5em + 0.5rem + calc(1px * 2))",
        padding: "0.25rem 0.5rem",
        fontSize: "0.875rem",
        borderRadius: "0.25rem",
      }
    : {};

const largeInputRule = ({ size }: RuleProps) =>
  size === "large"
    ? {
        minHeight: "calc(1.5em + 1rem + 2px)",
        padding: "0.5rem 1rem",
        fontSize: "1.25rem",
        borderRadius: "0.5rem",
      }
    : {};

const readOnlyPlaintextRule = ({ readOnly, plainText }: RuleProps) =>
  readOnly && plainText
    ? {
        display: "block",
        width: "100%",
        padding: "0.375rem 0",
        marginBottom: 0,
        lineHeight: 1.5,
        color: "#212529",
        backgroundColor: "transparent",
        border: "solid transparent",
        borderWidth: "1px 0",
        ":focus": {
          outline: 0,
        },
      }
    : {};

const readOnlyPlainTextSmallRule = ({
  readOnly,
  plainText,
  size,
}: RuleProps) =>
  readOnly && plainText && size === "small"
    ? {
        paddingRight: 0,
        paddingLeft: 0,
      }
    : {};

const readOnlyPlainTextLargeRule = ({
  readOnly,
  plainText,
  size,
}: RuleProps) =>
  readOnly && plainText && size === "large"
    ? {
        paddingRight: 0,
        paddingLeft: 0,
      }
    : {};

export type FloatingTextInputProps = Omit<
  TextInputProps,
  "readOnly" | "plainText" | "disabled"
>;

export const FloatingTextInput = ({
  className = "",
  id,
  size = "medium",
  value: controlledValue,
  type = "text",
  defaultValue,
  onChange,
  ...props
}: FloatingTextInputProps) => {
  const {
    readOnly = false,
    plainText = false,
    disabled = false,
    group = false,
    inputRef,
    setFocused,
    setPlaceholderShown,
  } = useFloatingTextContext() as FloatingTextContextValue;
  const { css } = useFela({ readOnly, plainText, disabled, size, group });
  const [uncontrolledValue, setUncontrolledValue] = useState(
    defaultValue ?? "",
  );

  const isControlled = controlledValue !== undefined;
  const value = isControlled ? controlledValue : uncontrolledValue;

  const cx = `${css([
    rule,
    groupRule,
    smallInputRule,
    smallGroupInputRule,
    largeInputRule,
    largeGroupInputRule,
    readOnlyPlaintextRule,
    readOnlyPlainTextSmallRule,
    readOnlyPlainTextLargeRule,
  ])} ${className}`.trim();

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (!isControlled) {
      setUncontrolledValue(e.target.value);
    }

    if (onChange) {
      onChange(e);
    }
  };

  const onFocus: FocusEventHandler<HTMLInputElement> = () => {
    if (
      !disabled &&
      !readOnly &&
      !plainText &&
      setFocused &&
      setPlaceholderShown
    ) {
      setFocused(true);
      setPlaceholderShown(false);
    }
  };

  const onBlur = (e: FocusEvent<HTMLInputElement>) => {
    if (setFocused && setPlaceholderShown) {
      setFocused(false);
      if (!e.target.value) {
        setPlaceholderShown(true);
      }
    }
  };

  return (
    <input
      data-testid="floating-text-input"
      type={type}
      ref={inputRef as ForwardedRef<HTMLInputElement>}
      className={cx}
      id={id}
      disabled={disabled}
      readOnly={readOnly}
      onFocus={onFocus}
      onBlur={onBlur}
      onChange={handleChange}
      value={value}
      {...props}
    />
  );
};
