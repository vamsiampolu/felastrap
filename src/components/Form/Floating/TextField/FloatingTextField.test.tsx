import { IRenderer } from "fela";
import { FloatingTextField } from "./FloatingTextField";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../../../test-helpers";
import { FloatingTextInput } from "./FloatingTextInput";
import { render, screen } from "@testing-library/react";

describe("components/Form/TextField/FloatingTextField", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });
  /*
  disabled = false,
  readOnly = false,
  plainText = false,
  */

  it("should render a floating text field with an input", async () => {
    const component = (
      <FloatingTextField id="foobar">
        <FloatingTextInput id="foobar-input" />
      </FloatingTextField>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("floating-text-field")).toBeInTheDocument();
    expect(screen.queryByTestId("floating-text-field")).toHaveClass(
      "form-floating",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a disabled floating text field with an input", async () => {
    const component = (
      <FloatingTextField disabled id="foobar">
        <FloatingTextInput id="foobar-input" />
      </FloatingTextField>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("floating-text-field")).toBeInTheDocument();
    expect(screen.queryByTestId("floating-text-field")).toHaveClass(
      "form-floating",
    );

    expect(screen.getByTestId("floating-text-input")).toBeInTheDocument();
    expect(screen.queryByTestId("floating-text-input")).toBeDisabled();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a readonly plaintext floating text field", async () => {
    const component = (
      <FloatingTextField readOnly plainText id="foobar">
        <FloatingTextInput id="foobar-input" />
      </FloatingTextField>
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("floating-text-field")).toBeInTheDocument();
    expect(screen.queryByTestId("floating-text-field")).toHaveClass(
      "form-floating",
    );

    expect(screen.getByTestId("floating-text-input")).toBeInTheDocument();
    expect(screen.queryByTestId("floating-text-input")).toHaveAttribute(
      "readonly",
    );

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
