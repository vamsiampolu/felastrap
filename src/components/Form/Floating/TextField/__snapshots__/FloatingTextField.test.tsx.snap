// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/TextField/FloatingTextField should render a disabled floating text field with an input 1`] = `
".a {
  display: block;
}
.b {
  position: relative;
}
.c {
  margin-bottom: 1rem !important;
}
.d {
  appearance: none;
}
.e {
  background-clip: padding-box;
}
.f {
  background-color: #fff;
}
.g {
  color: #212529;
}
.h {
  font-size: 1rem;
}
.i {
  font-weight: 400;
}
.j {
  height: calc(3.5rem + 2px);
}
.k {
  line-height: 1.25;
}
.l {
  min-height: calc(3.5rem + 2px);
}
.m {
  width: 100%;
}
.r:disabled {
  background-color: #e9ecef;
}
.s:disabled {
  opacity: 1;
}
.ac::placeholder {
  opacity: 1;
}
.ae::placeholder {
  color: transparent;
}
.af:not(:placeholder-shown) {
  padding-top: 1.625rem;
}
.ag:not(:placeholder-shown) {
  padding-bottom: 0.625rem;
}
.ah:-webkit-autofill {
  padding-top: 1.625rem;
}
.ai:-webkit-autofill {
  padding-bottom: 0.625rem;
}
.aj::-webkit-input-placeholder {
  opacity: 1;
}
.ak::-webkit-input-placeholder {
  color: transparent;
}
.al::-moz-placeholder {
  opacity: 1;
}
.am::-moz-placeholder {
  color: transparent;
}
.an:-ms-input-placeholder {
  opacity: 1;
}
.ao:-ms-input-placeholder {
  color: transparent;
}
.ap:-moz-placeholder {
  opacity: 1;
}
.aq:-moz-placeholder {
  color: transparent;
}
.ar {
  transition-property: border-color, box-shadow;
}
.as {
  transition-duration: 0.15s, 0.15s;
}
.at {
  transition-delay: ease-in-out, ease-in-out;
}
.au {
  transition-timing-function: ease, ease;
}
.av {
  padding-top: 1rem;
}
.aw {
  padding-right: 0.75rem;
}
.ax {
  padding-bottom: 1rem;
}
.ay {
  padding-left: 0.75rem;
}
.az {
  border-top-left-radius: 0.375rem;
}
.ba {
  border-top-right-radius: 0.375rem;
}
.bb {
  border-bottom-right-radius: 0.375rem;
}
.bc {
  border-bottom-left-radius: 0.375rem;
}
.bd {
  border-top-width: 1px;
}
.be {
  border-right-width: 1px;
}
.bf {
  border-bottom-width: 1px;
}
.bg {
  border-left-width: 1px;
}
.bh {
  border-top-style: solid;
}
.bi {
  border-right-style: solid;
}
.bj {
  border-bottom-style: solid;
}
.bk {
  border-left-style: solid;
}
.bl {
  border-top-color: #dee2e6;
}
.bm {
  border-right-color: #dee2e6;
}
.bn {
  border-bottom-color: #dee2e6;
}
.bo {
  border-left-color: #dee2e6;
}
.t:focus {
  background-color: #fff;
}
.u:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.v:focus {
  color: #212529;
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
@media (prefers-reduced-motion: reduce) {
  .n {
    transition-property: none;
  }
  .o {
    transition-duration: 0s;
  }
  .p {
    transition-delay: 0s;
  }
  .q {
    transition-timing-function: ease;
  }
}


<div data-testid="floating-text-field" class="a b c form-floating">
  <input
    data-testid="floating-text-input"
    type="text"
    class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb bc bd be bf bg bh bi bj bk bl bm bn bo d e f g h i j k l m n o p q r s t u v w x y z"
    id="foobar-input"
    disabled=""
    value=""
  />
</div>
;
"
`;

exports[`components/Form/TextField/FloatingTextField should render a floating text field with an input 1`] = `
".a {
  display: block;
}
.b {
  position: relative;
}
.c {
  margin-bottom: 1rem !important;
}
.d {
  appearance: none;
}
.e {
  background-clip: padding-box;
}
.f {
  background-color: #fff;
}
.g {
  color: #212529;
}
.h {
  font-size: 1rem;
}
.i {
  font-weight: 400;
}
.j {
  height: calc(3.5rem + 2px);
}
.k {
  line-height: 1.25;
}
.l {
  min-height: calc(3.5rem + 2px);
}
.m {
  width: 100%;
}
.r:disabled {
  background-color: #e9ecef;
}
.s:disabled {
  opacity: 1;
}
.ac::placeholder {
  opacity: 1;
}
.ae::placeholder {
  color: transparent;
}
.af:not(:placeholder-shown) {
  padding-top: 1.625rem;
}
.ag:not(:placeholder-shown) {
  padding-bottom: 0.625rem;
}
.ah:-webkit-autofill {
  padding-top: 1.625rem;
}
.ai:-webkit-autofill {
  padding-bottom: 0.625rem;
}
.aj::-webkit-input-placeholder {
  opacity: 1;
}
.ak::-webkit-input-placeholder {
  color: transparent;
}
.al::-moz-placeholder {
  opacity: 1;
}
.am::-moz-placeholder {
  color: transparent;
}
.an:-ms-input-placeholder {
  opacity: 1;
}
.ao:-ms-input-placeholder {
  color: transparent;
}
.ap:-moz-placeholder {
  opacity: 1;
}
.aq:-moz-placeholder {
  color: transparent;
}
.ar {
  transition-property: border-color, box-shadow;
}
.as {
  transition-duration: 0.15s, 0.15s;
}
.at {
  transition-delay: ease-in-out, ease-in-out;
}
.au {
  transition-timing-function: ease, ease;
}
.av {
  padding-top: 1rem;
}
.aw {
  padding-right: 0.75rem;
}
.ax {
  padding-bottom: 1rem;
}
.ay {
  padding-left: 0.75rem;
}
.az {
  border-top-left-radius: 0.375rem;
}
.ba {
  border-top-right-radius: 0.375rem;
}
.bb {
  border-bottom-right-radius: 0.375rem;
}
.bc {
  border-bottom-left-radius: 0.375rem;
}
.bd {
  border-top-width: 1px;
}
.be {
  border-right-width: 1px;
}
.bf {
  border-bottom-width: 1px;
}
.bg {
  border-left-width: 1px;
}
.bh {
  border-top-style: solid;
}
.bi {
  border-right-style: solid;
}
.bj {
  border-bottom-style: solid;
}
.bk {
  border-left-style: solid;
}
.bl {
  border-top-color: #dee2e6;
}
.bm {
  border-right-color: #dee2e6;
}
.bn {
  border-bottom-color: #dee2e6;
}
.bo {
  border-left-color: #dee2e6;
}
.t:focus {
  background-color: #fff;
}
.u:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.v:focus {
  color: #212529;
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
@media (prefers-reduced-motion: reduce) {
  .n {
    transition-property: none;
  }
  .o {
    transition-duration: 0s;
  }
  .p {
    transition-delay: 0s;
  }
  .q {
    transition-timing-function: ease;
  }
}


<div data-testid="floating-text-field" class="a b c form-floating">
  <input
    data-testid="floating-text-input"
    type="text"
    class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb bc bd be bf bg bh bi bj bk bl bm bn bo d e f g h i j k l m n o p q r s t u v w x y z"
    id="foobar-input"
    value=""
  />
</div>
;
"
`;

exports[`components/Form/TextField/FloatingTextField should render a readonly plaintext floating text field 1`] = `
".a {
  display: block;
}
.b {
  position: relative;
}
.c {
  margin-bottom: 1rem !important;
}
.d {
  appearance: none;
}
.e {
  background-clip: padding-box;
}
.f {
  background-color: transparent;
}
.g {
  color: #212529;
}
.h {
  font-size: 1rem;
}
.i {
  font-weight: 400;
}
.j {
  height: calc(3.5rem + 2px);
}
.k {
  line-height: 1.5;
}
.l {
  min-height: calc(3.5rem + 2px);
}
.m {
  width: 100%;
}
.r:disabled {
  background-color: #e9ecef;
}
.s:disabled {
  opacity: 1;
}
.ac::placeholder {
  opacity: 1;
}
.ae::placeholder {
  color: transparent;
}
.af:not(:placeholder-shown) {
  padding-top: 1.625rem;
}
.ag:not(:placeholder-shown) {
  padding-bottom: 0.625rem;
}
.ah:-webkit-autofill {
  padding-top: 1.625rem;
}
.ai:-webkit-autofill {
  padding-bottom: 0.625rem;
}
.aj {
  margin-bottom: 0;
}
.ak::-webkit-input-placeholder {
  opacity: 1;
}
.al::-webkit-input-placeholder {
  color: transparent;
}
.am::-moz-placeholder {
  opacity: 1;
}
.an::-moz-placeholder {
  color: transparent;
}
.ao:-ms-input-placeholder {
  opacity: 1;
}
.ap:-ms-input-placeholder {
  color: transparent;
}
.aq:-moz-placeholder {
  opacity: 1;
}
.ar:-moz-placeholder {
  color: transparent;
}
.as {
  border-top-width: 1px;
}
.at {
  border-right-width: 0;
}
.au {
  border-bottom-width: 1px;
}
.av {
  border-left-width: 0;
}
.aw {
  transition-property: border-color, box-shadow;
}
.ax {
  transition-duration: 0.15s, 0.15s;
}
.ay {
  transition-delay: ease-in-out, ease-in-out;
}
.az {
  transition-timing-function: ease, ease;
}
.ba {
  padding-top: 0.375rem;
}
.bb {
  padding-right: 0;
}
.bc {
  padding-bottom: 0.375rem;
}
.bd {
  padding-left: 0;
}
.be {
  border-top-left-radius: 0.375rem;
}
.bf {
  border-top-right-radius: 0.375rem;
}
.bg {
  border-bottom-right-radius: 0.375rem;
}
.bh {
  border-bottom-left-radius: 0.375rem;
}
.bi {
  border-top-style: solid;
}
.bj {
  border-right-style: solid;
}
.bk {
  border-bottom-style: solid;
}
.bl {
  border-left-style: solid;
}
.bm {
  border-top-color: transparent;
}
.bn {
  border-right-color: transparent;
}
.bo {
  border-bottom-color: transparent;
}
.bp {
  border-left-color: transparent;
}
.t:focus {
  background-color: #fff;
}
.u:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.v:focus {
  color: #212529;
}
.w:focus {
  outline-width: 0;
}
.x:focus {
  border-top-color: #86b7fe;
}
.y:focus {
  border-right-color: #86b7fe;
}
.z:focus {
  border-bottom-color: #86b7fe;
}
.ab:focus {
  border-left-color: #86b7fe;
}
@media (prefers-reduced-motion: reduce) {
  .n {
    transition-property: none;
  }
  .o {
    transition-duration: 0s;
  }
  .p {
    transition-delay: 0s;
  }
  .q {
    transition-timing-function: ease;
  }
}


<div data-testid="floating-text-field" class="a b c form-floating">
  <input
    data-testid="floating-text-input"
    type="text"
    class="a ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb bc bd be bf bg bh bi bj bk bl bm bn bo bp d e f g h i j k l m n o p q r s t u v w x y z"
    id="foobar-input"
    readonly=""
    value=""
  />
</div>
;
"
`;
